man/man1/bpan.1
.\" Automatically generated by Pod::Man 4.10 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "STDIN 1"
.TH "bpan" 1 "2023-01-01" "md2man v0.1.0" "Bash Package Manager"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "BPAN"
.IX Header "BPAN"
Bash Packages Are Nearby
.SS "\s-1BPAN DOCUMENTATION\s0"
.IX Subsection "BPAN DOCUMENTATION"
.IP "\(bu" 4
[[bpan\-overview]]
.IP "\(bu" 4
[[bpan\-user\-guide]]
.IP "\(bu" 4
[[bpan\-author\-guide]]
.SS "\s-1SYNOPSIS\s0"
.IX Subsection "SYNOPSIS"
Using the \f(CW\*(C`bpan\*(C'\fR \s-1CLI\s0 tool:
.PP
.Vb 5
\&    $ bpan \-\-help
\&    $ bpan show \-\-commands
\&    $ bpan help <command>
\&    $ bpan find foo
\&    $ bpan install somebody/foo\-bar\-bash
.Ve
.PP
Use the \f(CW\*(C`bpan\*(C'\fR Bash library in your Bash code:
.PP
.Vb 1
\&    #!/usr/bin/env bash
\&
\&    source "$MY_PROJECT_ROOT/.bpan/lib/bpan.bash" \-\-
\&
\&    bpan:source bashplus +err +sys
\&    bpan:source getopt
\&    bpan:source json
.Ve
.PP
Create new packages with \s-1BPAN:\s0
.PP
.Vb 5
\&    $ bpan init \-\-new=foo\-bar\-bash \-\-lib
\&    $ bpan add \-\-pkg=json\-bash
\&    $ bpan add \-\-file=test/foo.t
\&    $ bpan update
\&    $ bpan test
.Ve
.PP
Publish your Bash packages:
.PP
.Vb 2
\&    $ bpan publish \-\-register   # Add new package entry to Index
\&    $ bpan publish              # Update package\*(Aqs Index entry
.Ve
.SS "\s-1DESCRIPTION\s0"
.IX Subsection "DESCRIPTION"
\&\s-1BPAN\s0 is the \*(L"Missing Bash Package Manager\*(R".
.PP
It does 3 big things for you and your Bash code:
.IP "\(bu" 4
Find/install \s-1CLI\s0 apps/programs/libaries written in Bash
.IP "\(bu" 4
Install, manage, bundle Bash deps for a project or package
.IP "\(bu" 4
Publish Bash packages that you make
.SS "\s-1INSTALLATION\s0"
.IX Subsection "INSTALLATION"
Installing \s-1BPAN\s0 is simple and everything happens in a single directory on your host machine. The standard location is \f(CW\*(C`$HOME/.bpan/\*(C'\fR but you can change it to anything you like.
.PP
\&\s-1BPAN\s0 is written in \*(L"modern\*(R" Bash. It is built over several of the reusable libraries it makes available.
.PP
That said, you don't need to use Bash as your interactive shell to use \s-1BPAN.\s0 Bash just needs to be installed (and often is by default) because Bash is the \*(L"programming language\*(R" in this scenario.
.PP
\&\s-1BPAN\s0 works and is tested under the following interactive shells: \f(CW\*(C`bash\*(C'\fR, \f(CW\*(C`zsh\*(C'\fR, \f(CW\*(C`fish\*(C'\fR, \f(CW\*(C`ksh\*(C'\fR, \f(CW\*(C`mksh\*(C'\fR, \f(CW\*(C`tcsh\*(C'\fR, \f(CW\*(C`yash\*(C'\fR, \f(CW\*(C`ash\*(C'\fR, \f(CW\*(C`dash\*(C'\fR, \f(CW\*(C`mrsh\*(C'\fR, \f(CW\*(C`posh\*(C'\fR and \f(CW\*(C`sh\*(C'\fR.
.PP
\fIInstalling with the \s-1BPAN\s0 Installation Script\fR
.IX Subsection "Installing with the BPAN Installation Script"
.PP
You can run this command from any of the above shells to install \s-1BPAN:\s0
.PP
.Vb 1
\&    bash <(curl \-sL get.bpan.org)
.Ve
.PP
The script will try to identify your interactive shell type, and will choose an installation location of \f(CW\*(C`$HOME/.bpan/\*(C'\fR.
.PP
You can override these with the \f(CW\*(C`BPAN_SHELL\*(C'\fR and \f(CW\*(C`BPAN_ROOT\*(C'\fR environment variables like this:
.PP
.Vb 1
\&    BPAN_SHELL=abcsh BPAN_ROOT=/some/path bash <(curl \-sL get.bpan.org)
.Ve
.PP
\fIInstall \s-1BPAN\s0 from its Source Repository\fR
.IX Subsection "Install BPAN from its Source Repository"
.PP
\&\s-1NOTE:\s0 It is common to use \f(CW\*(C`$HOME/.bpan\*(C'\fR for \f(CW\*(C`/path/to/bpan\*(C'\fR in the following:
.IP "1. Clone the \s-1BPAN\s0 repo:" 4
.IX Item "1. Clone the BPAN repo:"
.Vb 1
\&    git clone https://github.com/bpan\-org/bpan /path/to/bpan
.Ve
.ie n .IP "1. Source the \s-1BPAN\s0 "".rc"" file to use ""bpan"" immediately. Use one of these commands, depending on your shell:" 4
.el .IP "1. Source the \s-1BPAN\s0 \f(CW.rc\fR file to use \f(CWbpan\fR immediately. Use one of these commands, depending on your shell:" 4
.IX Item "1. Source the BPAN .rc file to use bpan immediately. Use one of these commands, depending on your shell:"
.Vb 12
\&    ash  |  BPAN_ROOT=/path/to/bpan . /path/to/bpan/.rc
\&    bash |                     source /path/to/bpan/.rc
\&    dash |  BPAN_ROOT=/path/to/bpan . /path/to/bpan/.rc
\&    fish |                     source /path/to/bpan/.rc
\&    ksh  |  BPAN_ROOT=/path/to/bpan . /path/to/bpan/.rc
\&    mksh |  BPAN_ROOT=/path/to/bpan . /path/to/bpan/.rc
\&    mrsh | BPAN_ROOT=/path/to/bpan; . /path/to/bpan/.rc
\&    posh |  BPAN_ROOT=/path/to/bpan . /path/to/bpan/.rc
\&    sh   |  BPAN_ROOT=/path/to/bpan . /path/to/bpan/.rc
\&    tcsh |                     source /path/to/bpan/.rc
\&    yash |                          . /path/to/bpan/.rc 2>/dev/null
\&    zsh  |                     source /path/to/bpan/.rc
.Ve
.ie n .IP "1. Add the command in step 2 to your shell's startup file (""$HOME/.bashrc"", etc)." 4
.el .IP "1. Add the command in step 2 to your shell's startup file (\f(CW$HOME/.bashrc\fR, etc)." 4
.IX Item "1. Add the command in step 2 to your shell's startup file ($HOME/.bashrc, etc)."
.PP
\fIPrerequisites\fR
.IX Subsection "Prerequisites"
.PP
\&\s-1BPAN\s0 needs just a few very common things to work:
.IP "\(bu" 4
Bash 3.2+
.RS 4
.IP "\(bu" 4
The \s-1BPAN\s0 system is written in Bash (of course)
.RE
.RS 4
.RE
.IP "\(bu" 4
Git 2.9+
.RS 4
.IP "\(bu" 4
Required by \f(CW\*(C`bpan\*(C'\fR \s-1CLI\s0 and many \s-1BPAN\s0 packages, but...
.IP "\(bu" 4
Not required by programs using \s-1BPAN\s0 packages (that don't require it)
.RE
.RS 4
.RE
.IP "\(bu" 4
GitHub account
.RS 4
.IP "\(bu" 4
Only needed to publish to \s-1BPAN\s0 indexes hosted on GitHub
.IP "\(bu" 4
The default \s-1BPAN\s0 index is on GitHub
.RE
.RS 4
.RE
.SS "\s-1COPYRIGHT AND LICENSE\s0"
.IX Subsection "COPYRIGHT AND LICENSE"
Copyright 2022\-2023 by Ingy do\*:t Net
.PP
This is free software, licensed under:
.PP
The \s-1MIT\s0 (X11) License


Changes
[version "0.1.107"]
date = 2023-01-01T13:13:34Z
done = Fix bug with finding host from remote url

[version "0.1.106"]
date = 2022-12-30T17:49:14Z
done = Refactor -x code

[version "0.1.105"]
date = 2022-12-21T18:14:52Z
done = Doc updates
done = Refactor publish code for getting package config fields

[version "0.1.104"]
date = 2022-12-18T19:07:05Z
done = Refactor db.bash
done = Move lib/util/* to lib/
done = Move bpan command bash libraries under lib/cmd/

[version "0.1.103"]
date = 2022-12-18T09:29:04Z
done = Implement more of 'bpan show'
done = Add $PWD/bin to PATH in test/init template

[version "0.1.102"]
date = 2022-12-16T08:33:36Z
done = Fix 'update --self' to fetch tags
done = Fix bug in 'bpan update' when not in bpan related dir

[version "0.1.101"]
date = 2022-12-15T22:02:36Z
done = Change 'bpan cmds' to 'bpan show --cmds'
done = Update a test
done = Add optional 'package.summary'
done = Support package.tag being optional
done = Add package.type to package data
done = Fix another publish.bash bug
done = Fix variable scope bug in publish.bash
done = Support new meaning for host.*.source
done = Drop 'host.*.domain' from index
done = Update bashplus/time.bash for fixed timestamp

[version "0.1.100"]
date = 2022-12-15-08-00-14
done = Implement bpan:get-pkg-vars function
done = Support host.*.clone config option
done = Fix bug in db:get-package-domain-owner-name
done = Redo logic in get-package-domain-owner-name

[version "0.1.99"]
date = 2022-12-13-17-54-27
done = Fix bug in templates

[version "0.1.98"]
date = 2022-12-13-15-04-27
done = A couple small fixes to tests

[version "0.1.97"]
date = 2022-12-13T18:05:49
done = Refactor publishing

[version "0.1.96"]
date = Fri Dec  9 06:59:17 AM PST 2022
done = Changes to .bpan/config

[version "0.1.95"]
date = Thu Dec  8 02:16:14 PM PST 2022
done = Fix bug in test

[version "0.1.94"]
date = Thu Dec  8 11:00:20 AM PST 2022
done = Change 'bpan new xxx' to 'bpan init --new=xxx'
done = Move Meta to .bpan/config

[version "0.1.93"]
date = Tue Dec  6 05:20:53 PM PST 2022
done = Merge register and bump into publish

[version "0.1.92"]
date = Mon Dec  5 12:52:37 PM PST 2022
done = Change package config template

[version "0.1.91"]
date = Mon Dec  5 10:59:06 AM PST 2022
done = Make publish and register use plugins
done = Fix bug in update.bash calling install:main
done = Change package.repo to package.source
done = Update bpan.version for bpan update

[version "0.1.90"]
date = Sat Dec  3 11:42:42 AM PST 2022
done = Remove uninstall from getopt spec
done = Change uninstall to 'install --uninstall'
done = Change search and list to find and 'find --installed'

[version "0.1.89"]
date = Fri Dec  2 08:23:26 AM PST 2022
done = Fix search algorithm
done = Change config index.from to index.source

[version "0.1.88"]
date = Tue Nov 29 09:49:14 PM PST 2022
done = Simplify [index] entries in config files
done = Fix publish bug for author value
done = Update all index fields when publishing

[version "0.1.87"]
date = Mon Nov 28 07:37:59 AM PST 2022
done = Fix bug when installing older package version
done = Fix bugs in 'bpan publish'

[version "0.1.86"]
date = Sun Nov 27 12:32:05 PM PST 2022
done = Use 'install_dir' for BPAN_INSTALL internally
done = Update global and local bpan.version for bump on bpan itself
done = Cache online resources so tests run faster and also offline
done = Make config file locations configurable
done = Support multiple BPAN indexes
done = Change 'bpan upgrade' to 'bpan update --self'

[version "0.1.85"]
date = Tue Nov 22 03:10:25 PM PST 2022
done = Fix bug in publish.bash

[version "0.1.84"]
date = Sun Nov 20 12:39:08 PM PST 2022
done = Correct paths used by bpan:source
done = Add 'bpan + cmd' misc command support

[version "0.1.83"]
date = Sat Nov 19 02:18:46 PM PST 2022
done = Small refactor in etc/config
done = Make upgrade test be branch aware
done = Remove api-version code
done = Add bashplus/str
done = Template syntax from '(%...%)' to '[%...%]'

[version "0.1.82"]
date = Fri Nov 18 07:59:02 AM PST 2022
done = Changes to make tests pass in older environments

[version "0.1.81"]
date = Fri Nov 18 07:20:55 AM PST 2022
done = Various changes for portability

[version "0.1.80"]
date = Thu Nov 17 11:07:17 PM PST 2022
done = Conversion to new index format

[version "0.1.79"]
date = Fri Nov 11 09:56:03 AM PST 2022
done = Switch to new bashplus +module syntax

[version "0.1.78"]
date = Fri Nov 11 08:14:43 AM PST 2022
done = Update index always for 'bpan update' command
done = Update bpan dependencies

[version "0.1.77"]
date = Tue Nov  8 10:03:47 AM PST 2022
done = Don't allow 'bpan publish' on bpan repo itself
done = Fix forced index updates
done = Don't check if file already exists to update it

[version "0.1.76"]
date = Tue Nov  8 08:47:12 AM PST 2022
done = Update test-tap
done = Get package host from config file

[version "0.1.75"]
date = Mon Nov  7 06:37:09 AM PST 2022
done = Move more hardcoded values into config

[version "0.1.74"]
date = Sun Nov  6 08:29:26 PM PST 2022
done = Support 'update.package = name:@/part'
done = Configuration file refactoring
done = Speed up testing
done = Fixes for 'bpan add --from=...'
done = Support BPAN_DEBUG_BASH_X

[version "0.1.73"]
date = Sat Nov  5 09:05:30 AM PDT 2022
done = Changes to make testing more reliable

[version "0.1.72"]
date = Thu Nov  3 01:21:40 PM PDT 2022
done = Fixes for .rc file and MANPATH
done = Revert "Remove use of non-portable 'manpath' command"
done = Make sure update is using correct config

[version "0.1.71"]
date = Thu Nov  3 07:59:56 AM PDT 2022
done = Renumber tests
done = Don't make unnecessary external bpan calls
done = Update bashplus with stack-trace support

[version "0.1.70"]
date = Wed Nov  2 11:10:21 AM PDT 2022
done = Various refactoring and updates
done = Support BPAN_BASH_DEBUG=1 for set -x
done = Make commit quiet for --new
done = Don't init/update man pages when pandoc 2 not installed
done = Make tests pass when pandoc 2+ not available

bpan.rb
Bundler.require
%w{
  sinatra
  json
  logger
  fileutils
  erb
  yaml
}.each{|l| require l}

$stdout.sync = true
logger = Logger.new($stdout, Logger::DEBUG)

def XXX *args
  c = caller_locations(1,1)[0].label
  l = caller_locations(1,1)[0].lineno
  args = args.first if args.size == 1
  logger.info "\nXXX from #{c} #{l}:\n#{YAML.dump args}"
end

post '/?' do
  request.body.rewind  # in case someone already read it
  data = body_json

  if data['action'] == 'started'
    sender = data['sender']
    logger.debug "Received star from #{sender['login'].inspect}"
    add_author sender
    return "Thanks for starring, #{sender['login']}"
  elsif data['ref_type'] == 'tag'
    created = !data['master_branch'].nil?
    logger.debug "Tag %s %s on %s" % [
      data['ref'].inspect,
      (created ? 'created' : 'deleted'),
      data['repository']['clone_url'].inspect,
    ]
    return 'Deleted tag' unless created
    meta = add_package data
    return "Thanks for pushing %s version %s sha %s" % [
      meta['name'].inspect,
      meta['version'].inspect,
      meta['release']['sha'],
    ]
  else
    logger.info data
    halt 400, "Invalid action"
  end
end

post '/rebuild/' do
  author = JSON.parse(File.read(AUTHOR_FILE))
  post_process_and_commit_author author
  msg = "Successful rebuild of indexes and webpage"
  logger.info msg
  return "#{msg}\n"
end

if ENV['RACK_ENV'] == 'development'
  get '/?' do
    File.open(AUTHOR_FILE, 'w+') {|f|f.puts'[]'} \
      unless File.exist?(AUTHOR_FILE)
    author = JSON.parse(File.read(AUTHOR_FILE))
    homepage(author)
  end
else
  get '/?' do
    redirect to('http://bpan.org')
  end
end

if %w{development test}.include? ENV['RACK_ENV']
  get '/css/jumbotron-narrow.css' do
    content_type 'text/css'
    File.read(
      File.join(
        File.dirname(__FILE__),
        'views',
        'jumbotron-narrow.css',
      )
    )
  end

  get '/author.json' do
    File.open(AUTHOR_FILE, 'w+') {|f|f.puts'[]'} \
      unless File.exist?(AUTHOR_FILE)
    return File.read(AUTHOR_FILE)
  end

  get '/package.json' do
    File.open(PACKAGE_FILE, 'w+') {|f|f.puts'{}'} \
      unless File.exist?(PACKAGE_FILE)
    return File.read(PACKAGE_FILE)
  end
end

#------------------------------------------------------------------------------
private

def h s
  CGI.escape_html s
end

GH_PAGES_BRANCH = 'gh-pages'
GH_PAGES_DIR = File.expand_path(
  File.join(
    File.dirname(__FILE__),
    ENV['BPAN_TEST_DIR'] || '',
    GH_PAGES_BRANCH,
  )
)
GIT_REMOTE =
  ENV['BPAN_TEST_REMOTE'] ||
  'git@github.com:bpan-org/bpan.git'

if not Dir.exist?(GH_PAGES_DIR)
  Git.clone(
    GIT_REMOTE,
    GH_PAGES_BRANCH,
    path: File.dirname(GH_PAGES_DIR),
    log: logger,
  )
  fail unless Dir.exist?(GH_PAGES_DIR)
end

GIT = Git.open GH_PAGES_DIR, log: logger
GIT.checkout(GH_PAGES_BRANCH)
GIT.pull 'origin', GH_PAGES_BRANCH

def ghpages_file *parts
  File.join(
    GH_PAGES_DIR,
    *parts,
  )
end

def ghpages_index_file file_name
  ghpages_file 'index', file_name
end

AUTHOR_FILE = ghpages_index_file 'author.json'
AUTHOR_FILEP = "#{AUTHOR_FILE}p"
PACKAGE_FILE = ghpages_index_file 'package.json'
PACKAGE_FILEP = "#{PACKAGE_FILE}p"
HOMEPAGE_VIEW = ERB.new(
  File.read File.join ghpages_file, '_cogweb', 'view', 'index.html.erb'
)
HOMEPAGE_FILE = ghpages_file 'index.html'

def add_author sender
  File.open(AUTHOR_FILE, 'w') {|f|f.puts'[]'} \
    unless File.exist?(AUTHOR_FILE)
  author = JSON.parse(File.read(AUTHOR_FILE))
  author << format_author(sender)
  post_process_and_commit_author author, sender
end

def post_process_and_commit_author author, hash=nil
  author.uniq! {|a| a["login"]}
  author.sort! {|a,b| a["login"] <=> b["login"]}

  # Regenerate homepage
  `(cd gh-pages/_cogweb; make site)`
  GIT.add(File.basename(HOMEPAGE_FILE))

  # Regenerate json index
  json = JSON.pretty_generate author
  File.open(AUTHOR_FILE, 'w') {|f|
    f.write json
  }
  File.open(AUTHOR_FILEP, 'w') {|f|
    f.write "var author = "
    f.write json
    f.write ";"
  }
  GIT.add(AUTHOR_FILE)
  GIT.add(AUTHOR_FILEP)
  begin
    message = if hash
      "[BPAN] Add author #{hash['login'].inspect}"
    else
      "[BPAN] Prune/resync author directory"
    end
    GIT.commit(message)
    GIT.push 'origin', GH_PAGES_BRANCH
  rescue Git::GitExecuteError => e
    raise e unless e.message =~ /nothing to commit/i
  end

  package = load_package
  save_package package
end

def add_package data
  meta = nil
  Dir.mktmpdir('bpan_package_clone') do |dir|
    url = data['repository']['clone_url']
    tag = data['ref']
    git = Git.clone(url, tag, path: dir, log: logger, depth: 0)
    git.checkout(tag)
    git.pull 'origin', tag
    meta = YAML.safe_load(File.read(File.join(dir, tag, 'Meta')))
    meta['release'] = {
      'sha' => git.revparse(tag),
      'url' => url,
      'timestamp' => Time.now.to_i
    }
  end

  # Get the existing package index
  package = load_package

  # generate our new entries
  half_qualified_key = [
    meta['name'],
    data['repository']['owner']['login'],
  ].join('/')
  fully_qualified_key = [
    half_qualified_key,
    meta['version'],
  ].join('/')

  # update version array
  package[half_qualified_key] ||= []
  package[half_qualified_key] = [package[half_qualified_key]] \
    if package[half_qualified_key].instance_of? String
  package[half_qualified_key].unshift meta['version']

  # add full package meta for this version
  package[fully_qualified_key] = meta

  # set short name if it isn't already taken
  package[meta['name']] ||= half_qualified_key

  save_package package, fully_qualified_key

  return meta
end

def load_package
  File.open(PACKAGE_FILE, 'w+') {|f|f.puts'{}'} \
    unless File.exist?(PACKAGE_FILE)
  JSON.parse(
    File.open(PACKAGE_FILE, "r:UTF-8", &:read)
  )
end

def save_package package, fully_qualified_key=nil
  # Regenerate json index. Pretty printed and top keys sorted.
  json = JSON.pretty_generate Hash[
    package.keys.sort.map{|k|[k,package[k]]}
  ]
  File.open(PACKAGE_FILE, 'w') {|f|
    f.write json
  }
  File.open(PACKAGE_FILEP, 'w') {|f|
    f.write "var package = "
    f.write json
    f.write ";"
  }
  GIT.add(PACKAGE_FILE)
  GIT.add(PACKAGE_FILEP)
  begin
    if fully_qualified_key
      GIT.commit("[BPAN] Add package #{fully_qualified_key.inspect}")
    else
      logger.debug "Regenerating package.json"
    end
  rescue Git::GitExecuteError => e
    raise e unless e.message =~ /nothing to commit/i
  end
  GIT.push 'origin', GH_PAGES_BRANCH
end

def homepage author
  HOMEPAGE_VIEW.result(binding)
end

def format_author author
  { "login" => author['login'],
    "avatar_url" => author["avatar_url"],
    "gravatar_id" => author["gravatar_id"],
    "url" => author["url"],
    "html_url" => author["html_url"],
  }
end

def body_json
  j = request.body.read
  # logger.info j
  JSON.parse j
rescue
  halt 400, "Invalid JSON"
end
